incubator/hnc/test/conversion/conversion_test.go

```diff
@@ -0,0 +1,144 @@
+package conversion
+
+import (
+	"os"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+
+	. "sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils"
+)
+
+const namspacePrefix = "e2e-conversion-test-"
+
+func TestConversion(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	SetDefaultEventuallyTimeout(time.Second * 2)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"HNC v1alpha2 Conversion Test Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = Describe("Conversion from v1alpha1 to v1alpha2", func() {
+	const (
+		// hncVersion has to be a version using v1alpha1.
+		hncVersion = "v0.5.0"
+		nsA    = namspacePrefix + "a"
+		nsB    = namspacePrefix + "b"
+		nsC    = namspacePrefix + "c"
+	)
+
+	var (
+		crds = []string{"subnamespaceanchors.hnc.x-k8s.io","hierarchyconfigurations.hnc.x-k8s.io", "hncconfigurations.hnc.x-k8s.io"}
+	)
+
+	BeforeEach(func() {
+		// We need GOPATH here to install kubectl-hns plugin.
+		CheckGoPath()
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Tear down HNC for both the specified version and at HEAD.
+		TearDownHNC(hncVersion)
+		// Wait for HNC to be fully torn down.
+		ensureHNCCleared(crds)
+	})
+
+	AfterEach(func() {
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Only tear down HNC at HEAD, since that's what we just deployed.
+		TearDownHNC("")
+	})
+
+	It("should convert CRD versions properly and HNC should still work", func() {
+		// Set up v1alpha1 and verify everything is fine before conversion.
+		GinkgoT().Log("Set up v1alpha1")
+		setupV1alpha1(hncVersion)
+		// Wait for the validating webhooks to be ready.
+		ensureVWHReady()
+		// Create namespace 'a' and a subnamespace 'b'
+		GinkgoT().Log("Apply v1alpha1 resources")
+		MustRun("kubectl apply -f v1alpha1_resources.yaml")
+		// Create a namespace 'c' and use kubectl plugin to set its parent to 'a'
+		MustRun("kubectl create ns", nsC)
+		GinkgoT().Log("Use v1alpha1 kubectl-hns plugin")
+		MustRun("kubectl hns set", nsC, "--parent", nsA)
+
+		// Verify kubectl-hns plugin works and 'b' and 'c' are children of 'a'.
```

Comment:
This should be a different test.

`It("should convert subnamespace anchors", ...)`
`It("should convert HCs with parents", ...)`
`It("should convert HCs with allowCascadingDelete", ...)`

etc.

BTW are we changing `allowCascadingDelete` to `allowCascadingDeletion`?

The only time I'd try to do something more end-to-end like is to make sure that conditions look good before and after. But I wouldn't test conditions exhaustively - I'd just make sure that each of the three objects can be converted in _one_ bad state each.

Reply:
Plus, as mentioned above, I wouldn't use `kubectl-hns` - I'd rather see the API objects directly so I can see what's getting converted.